contracts/staking-vault/StakingVault.sol

```diff
@@ -0,0 +1,248 @@
+pragma solidity ^0.8.0;
+
+import "@openzeppelin-4.7.0/token/ERC1155/ERC1155.sol";
+import "@openzeppelin-4.7.0/token/ERC20/utils/SafeERC20.sol";
+import "@openzeppelin-4.7.0/token/ERC20/extensions/IERC20Metadata.sol";
+
+import "../../interfaces/IPriceFeeds.sol";
+import "../proxies/0_8/Upgradeable_0_8.sol";
+contract StakingVault is Upgradeable_0_8, ERC1155 {
+    using SafeERC20 for IERC20;
+    struct tokenComposition {
+        address depositToken;
+        address tokenToBack;
+    }
+    address public protocol;
+    IPriceFeeds public priceFeed;
+    address public valuationToken;
+    address[] public stakingTokens;
+    mapping(address=>bool) public tokenSupported;
+    mapping(uint256=>uint256) public sTokenPrice;
+    mapping(uint256=>tokenComposition) public IDToTokens;
+    mapping(uint256=>uint256) public balanceStakedPerID;
+    mapping(uint256=>uint256) public supplyPerID;
+
+    mapping(address=>uint256) public undistributedRewards;
+    mapping(uint256=>uint256) public rewardsPerToken;
+    address public rewardToken;
+    mapping(address=>mapping(uint256=>uint256)) public lastClaimRewardAccrual;
+    mapping(uint256=>bool) public initialized;
+
+    event PriceFeed(address feed);
+    event Protocol(address protocol);
+    event ValuationToken(address token);
+    event TokensSupported(address[] supportedTokens);
+    event Deposit(address depositor, address tokenBacked, address depositToken, uint256 amount);
+    event Withdraw(address withdrawer, address tokenBacked, address receivedToken, uint256 amount);
+    event RewardToken(address token);
+    event PoolDraw(address tokenBacked, address tokenCovered, uint256 amountCovered);
+    event AddRewards(address tokenBacked, uint256 amount);
+    event AccumulateRewards(address tokenBacked, uint256 amount);
+    event DistributeRewards(address tokenBacked, uint256 amount);
+
+    modifier onlyProtocol() {
+        require(msg.sender==protocol, "not protocol");_;
+    }
+
+    constructor(string memory uri_)ERC1155(uri_) {}
+
+    function updateTokenSupport(address[] memory tokens, bool[] memory support, address[] memory listOfSupportedTokens) external onlyOwner {
```

Comment:
is `tokens` and `listOfSupportedTokens` going to have the same value? why not pass just one and then loop thru it?

Reply:
no, tokens may include tokens to be removed if support is set to false and listOfSupportedTokens is the full list of all tokens supported. the mapping is used to reduce cost of checking if token is supported but entire list of staking tokens is needed as well.